### PR TITLE
Optimize model execution with CUDA graph

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR uses CUDA graph to optimize the CPU overheads in model execution. This is particularly effective for small models and when using tensor parallelism.

Many thanks to @scv119 and @Yard1 for addressing the NCCL-related issues.

### CODE REVIEW

**Comment by casper-hansen** [comment]
Great to see we can optimize model execution with CUDA graphs! How much of a boost in throughput is this for 7B models - perhaps with TP=1 and TP=8?

**Comment by hanzhi713** [comment]
@WoosukKwon  Nice change! Once this is merged, I will upstream my custom allreduce kernels

See this doc: [Latency-optimal allreduce and cuda graph optimization.pdf](https://github.com/vllm-project/vllm/files/13574639/Latency-optimal.allreduce.and.cuda.graph.optimization.pdf)

**Review by zhuohan123 — COMMENTED** [review]
Thanks for the contribution! Left comments on the problems I found. BTW, is there any significant memory overhead brought by cuda graph that affects the KV cache memory profiling?

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -332,6 +527,16 @@ def _make_tensor_with_pad(
...
+    return torch.tensor(padded_x, dtype=dtype, device=device)
+
+
+def _get_graph_batch_size(batch_size: int) -> int:
+    if batch_size <= 2:
+        return batch_size
+    elif batch_size <= 4:
+        return 4
+    else:
+        return (batch_size + 7) // 8 * 8
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:542]: Yeah that was actually my initial implementation, but I was a bit worried about the performance. Given that the logic won't change much, why don't we just add a comment to `_BATCH_SIZES_TO_CAPTURE`?

File: requirements.txt [hunk 2]
```diff
@@ -13,3 +13,4 @@ fastapi
 uvicorn[standard]
 pydantic == 1.10.13  # Required for OpenAI server.
 aioprometheus[starlette]
+cupy-cuda12x  # Required for CUDA graphs. CUDA 11.8 users should install cupy-cuda11x instead.
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply requirements.txt:16]: Yeah, that's a good point. Now that we have quite complex dependencies, we need to discard `requirements.txt` and handle them by code in `setup.py`.

File: vllm/model_executor/parallel_utils/cupy_utils.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        f"{rank=} should be a integer between [0, {world_size})"
+    cupy.cuda.runtime.setDevice(torch.cuda.current_device())
+    _NCCL_BACKEND = NCCLBackendWithBFloat16(world_size, rank, host, port)
+    _WORLD_SIZE = world_size
+
+
+def all_reduce(input_: torch.Tensor, op=ReduceOp.SUM) -> None:
+    """All-reduces the input tensor across the process group."""
+    assert input_.is_cuda,\
+        f"{input_} should be a cuda tensor"
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/parallel_utils/cupy_utils.py:81]: Fixed

File: vllm/worker/model_runner.py [hunk 4]
```diff
@@ -321,6 +381,141 @@ def profile_run(self) -> None:
...
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[Tuple[torch.Tensor, torch.Tensor]],
+        input_metadata: InputMetadata,
+    ) -> torch.Tensor:
+        # KV caches are fixed tensors, so we don't need to copy them.
+        assert kv_caches == self.input_buffers["kv_caches"]
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: Good point. Removed.

File: vllm/worker/model_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:573]: Added a comment

File: vllm/model_executor/parallel_utils/cupy_utils.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            "NCCLBackend is not available. Please install cupy.") from cupy
+
+    global _NCCL_BACKEND
+    global _WORLD_SIZE
+    assert world_size > 0, f"{world_size=} should be a positive integer"
+    assert 0 <= rank < world_size,\
+        f"{rank=} should be a integer between [0, {world_size})"
+    cupy.cuda.runtime.setDevice(torch.cuda.current_device())
+    _NCCL_BACKEND = NCCLBackendWithBFloat16(world_size, rank, host, port)
+    _WORLD_SIZE = world_size
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/parallel_utils/cupy_utils.py]: Added!

**Comment by zhisbug** [comment]
Per my tests:
1. big memory overhead once cuda graph is on
2. api_endpoint entrypoint does not work when I turned cuda graph on.

File: vllm/worker/model_runner.py [hunk 6]
```diff
@@ -154,34 +176,69 @@ def _prepare_decode(
...
                                     dtype=torch.int,
-                                    device="cuda")
-        max_block_table_len = max([len(t) for t in block_tables])
-        block_tables = _make_tensor_with_pad(block_tables,
-                                             max_len=max_block_table_len,
-                                             pad=0,
-                                             dtype=torch.int)
+                                    device=device)
+
+        # FIXME
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: Removed!

File: vllm/worker/model_runner.py [hunk 7]
```diff
@@ -154,34 +176,69 @@ def _prepare_decode(
...
+        if max_context_len <= self.max_context_len_to_capture:
+            input_block_tables = self.block_tables[:batch_size]
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: Added more detailed comments on `graph_block_tables` under `__init__`.

File: vllm/worker/worker.py [hunk 8]
```diff
@@ -174,8 +191,22 @@ def _init_distributed_environment(
...
+        if cupy_world_size != parallel_config.world_size:
+            raise RuntimeError(
+                "cupy.distributed is already initialized but the cupy world "
+                "size does not match parallel_config.world_size "
+                f"({cupy_world_size} vs. {parallel_config.world_size}).")
+    else:
+        cupy_utils.init_process_group(
+            world_size=parallel_config.world_size,
+            rank=rank,
+        )
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/worker.py]: As we discussed offline, I deferred this as TODO and left a comment.

File: requirements.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply requirements.txt]: Can we do it in another PR? I think we need a PR just for cleaning up the dependencies.

**Comment by sfc-gh-hazhang** [comment]
btw, 2 is because of the increased memory overhead. I have to turn down GPU utilization value even more to make it work in online

File: vllm/worker/model_runner.py [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
 logger = init_logger(__name__)
 
+KVCache = Tuple[torch.Tensor, torch.Tensor]
 _PAD_SLOT_ID = -1
+# Capture graphs for batch size 1, 2, 4, 8, 16, 24, 32, 40, ..., 256.
+# NOTE: _get_graph_batch_size needs to be updated if this list is changed.
+_BATCH_SIZES_TO_CAPTURE = [1, 2, 4] + [8 * i for i in range(1, 33)]
+# If the context length of a sequence is larger than this, we fall back to the
+# eager mode.
+_MAX_CONTEXT_LEN_TO_CAPTURE = 8192
```
Comment (Yard1) [inline vllm/worker/model_runner.py:25]: should this be configurable?

**Comment by zhisbug** [comment]
Which recent commit corresponds to fixing the memory issue?

**Comment by WoosukKwon** [comment]
@zhisbug Actually, we recently changed the memory profiling method in #2031. Now it should be more conservative in distributed setting.

**Comment by WoosukKwon** [comment]
@zhisbug 
> api_endpoint entrypoint does not work when I turned cuda graph on.

I'm not able to reproduce this. Could you provide a reproducible script?

**Comment by zhisbug** [comment]
> @zhisbug 
> 
> > api_endpoint entrypoint does not work when I turned cuda graph on.
> 
> 
> 
> I'm not able to reproduce this. Could you provide a reproducible script?

I think good to merge. 2 is because of OOM. Your new profiling should address it as well.

File: vllm/worker/model_runner.py (hunk 9 above)
Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: Good point. Added an argument `max_context_len_to_capture`.

**Comment by WoosukKwon** [comment]
@zhuohan123 I've addressed your comments. PTAL.

File: vllm/worker/model_runner.py [hunk 10]
```diff
@@ -319,8 +387,139 @@ def profile_run(self) -> None:
...
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[Tuple[torch.Tensor, torch.Tensor]],
+        input_metadata: InputMetadata,
+    ) -> torch.Tensor:
+        # KV caches are fixed tensors, so we don't need to copy them.
+        del kv_caches
+
+        # Copy the input tensors to the input buffers.
+        self.input_buffers["input_ids"].copy_(input_ids)
```
Comment (scv119) [inline vllm/worker/model_runner.py:508]: to save memory, you can share the input_buffers across all graphs.

File: vllm/worker/model_runner.py [hunk 11]
```diff
@@ -32,12 +41,31 @@ def __init__(
         self.model = None
         self.block_size = None  # Set after initial profiling.
 
+        self.graph_runners: Dict[int, CUDAGraphRunner] = {}
```
Comment (scv119) [inline vllm/worker/model_runner.py:44]: another way to save memory is to use LRUcache for this map, and capture it on demand.

**Review by scv119 — COMMENTED** [review]
we've also made pad size (other than 8) configurable to save memory.

File: vllm/worker/model_runner.py (hunk 10 above)
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:508]: Hi @scv119, I actually reuse the input buffers used for capturing all graphs. https://github.com/vllm-project/vllm/blob/13e4e762ae6b3d2f89b69a2b571ccb7ad455f496/vllm/worker/model_runner.py#L402

**Review by zhuohan123 — APPROVED** [review]
In general LGTM! Thanks for the great work! Left some minor comments.

File: vllm/engine/arg_utils.py [hunk 12]
```diff
@@ -172,6 +174,12 @@ def add_cli_args(
                             type=int,
                             default=EngineArgs.max_paddings,
                             help='maximum number of paddings in a batch')
+        parser.add_argument('--max-context-len-to-capture',
+                            type=int,
+                            default=EngineArgs.max_context_len_to_capture,
+                            help='maximum context length covered by CUDA '
+                            'graphs. When a sequence has context length '
+                            'larger than this, we fall back to eager mode.')
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/engine/arg_utils.py:182]: Sounds good. Fixed.

File: vllm/config.py [hunk 13]
```diff
@@ -49,6 +49,9 @@ class ModelConfig:
             output). If None, will be derived from the model.
         quantization: Quantization method that was used to quantize the model
             weights. If None, we assume the model weights are not quantized.
+        enforce_eager: Whether to enforce eager execution. If True, we will
+            disable CUDA graph and always execute the model in eager mode.
+            If False, we will use CUDA graph and eager execution in hybrid.
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/config.py:54]: Added. Thanks!

File: vllm/model_executor/input_metadata.py [hunk 14]
```diff
@@ -21,12 +21,14 @@ def __init__(
         max_context_len: Optional[int],
         context_lens: Optional[torch.Tensor],
         block_tables: Optional[torch.Tensor],
+        use_graph: bool,
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/input_metadata.py]: Good point! Fixed!

File: vllm/entrypoints/llm.py [hunk 15]
```diff
@@ -71,6 +74,7 @@ def __init__(
         seed: int = 0,
         gpu_memory_utilization: float = 0.9,
         swap_space: int = 4,
+        enforce_eager: bool = False,
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/entrypoints/llm.py:81]: I'm neutral on this. It's a bit too detailed, but some users may find it useful.

File: vllm/worker/model_runner.py (hunk 11 above)
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:44]: Made a new issue on this. Let's do it in a followup PR.

File: vllm/entrypoints/llm.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/entrypoints/llm.py:81]: Hmm.. Let's add it. It doesn't hurt us.

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/entrypoints/llm.py:81]: Hmm.. Let's add it. It doesn't hurt us.

**Comment by WoosukKwon** [comment]
Hi @scv119 @Yard1, many thanks for your review and all the support. I've added you two as the co-authors of this PR. 😄

**Comment by WoosukKwon** [comment]
@hanzhi713 We've just merged the PR. Looking forward to your PR!

**Comment by esmeetu** [comment]
FYI. Putting my experiment here:
Running a 34b model with TP8 will take up more than 2GB memory on every device than eager mode.
And There is no speedup on T4 GPU with batch 1. It seems cuda graph not nice for older arch gpu cards.

**Comment by corbt** [comment]
How large is the expected speedup here? I have a stress test where I hit vLLM with 50 simultaneous inference requests against a Mistral 7B model that have 800 input tokens and 200 output tokens. I run for 150 seconds and count the number of requests that complete successfully during that time. I'm seeing no measurable performance difference between vLLM 0.2.0 (which we run in prod) and 0.2.6. In both cases we average about 260 requests over the 150-second period.

**Comment by lionsky** [comment]
> This PR uses CUDA graph to optimize the CPU overheads in model execution. This is particularly effective for small models and when using tensor parallelism.

wonder if there is some performance data?

**Pull request merged**